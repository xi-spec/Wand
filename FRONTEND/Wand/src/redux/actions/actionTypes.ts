export default {
  LOAD_PRODUCT_TYPE: 'LOAD_PRODUCT_TYPE',
  LOAD_ALL_PRODUCTS: 'LOAD_ALL_PRODUCTS',
  LOAD_NEW_PRODUCTS: 'LOAD_NEW_PRODUCTS',
  LOAD_PRODUCTS_LIST: 'LOAD_PRODUCTS_LIST',
  LOAD_PRODUCT_DETAIL: 'LOAD_PRODUCT_DETAIL',
  ADD_PRODUCT_TO_CART: ' ADD_PRODUCT_TO_CART',
  INCREASE_QUANTITY_CART: 'INCREASE_QUANTITY_CART',
  DECREASE_QUANTITY_CART: 'DECREASE_QUANTITY_CART',
  DELETE_PRODUCT_CART: 'DELETE_PRODUCT_CART',
  DELETE_ALL_CART: 'DELETE_ALL_CART',
  UPDATE_STOCK_API: 'UPDATE_STOCK_API',
  SEARCH_PRODUCT: 'SEARCH_PRODUCT',
  ADD_PRODUCT_TO_FAVOURITE: 'ADD_PRODUCT_TO_FAVOURITE',
  REGISTER: 'REGISTER',
  REGISTER_ERROR: 'REGISTER_ERROR',
  LOGIN: 'LOGIN',
  LOGIN_ERROR: 'LOGIN_ERROR',
  CLEAN_ISEXIST_STATUS: 'CLEAN_STATUS',
  CLEAN_LOGIN_ERROR: 'CLEAN_LOGIN_ERROR',
  CLEAN_REGISTED: 'CLEAN_REGISTED',
  REMOVE_FAVOURITE: 'REMOVE_FAVOURITE',
  LOGOUT: 'LOGOUT',
  UPDATE_USER: 'UPDATE_USER',
  PAYMENT_SUCCSSFULY: 'PAYMENT_SUCCESSFULY',
  PDATE_STOCK_API: 'PDATE_STOCK_API'
}
;
