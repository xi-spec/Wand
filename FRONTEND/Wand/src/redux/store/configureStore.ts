import reduxImmutableStatInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import { applyMiddleware, compose, createStore } from 'redux';
import rootReducer from '../reducers';

function configureStore () {
  const composeEnhancers = compose;

  return createStore(
    rootReducer,
    composeEnhancers(applyMiddleware(reduxImmutableStatInvariant(), thunk))
  );
}

export default configureStore();
