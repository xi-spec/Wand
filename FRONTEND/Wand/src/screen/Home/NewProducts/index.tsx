import React from 'react';
import {
  Text,
  TouchableOpacity,
  View,
  ScrollView,
  Image
} from 'react-native';
import { connect } from 'react-redux';
import styles from './styles';
import { NavigationScreenProp, NavigationRoute } from 'react-navigation';
import { Params } from '../../../models';
import { Product } from '../../../models/initiaState';
import setCostValue2Decimals from '../../../utils/setCostValue2Decimals';

function NewProducts ({ allProducts, navigation }:{
  navigation: NavigationScreenProp<NavigationRoute<Params>, Params>, allProducts:Product[]}) {
  const randNum = Math.floor(Math.random() * 14);
  const newPrdocuts = allProducts.slice(randNum, randNum + 6);

  return (
    <View style={styles.container}>
   <Text style={styles.text}>
     New arrivals
   </Text>
    <ScrollView
    bounces={false}
    horizontal={true}
    showsHorizontalScrollIndicator={false} >
   {newPrdocuts && newPrdocuts.map((product) => (
     <View
     key={Math.random() * 10}>

    <TouchableOpacity
    style={styles.imageContainer}

    testID='newProduts'
    onPress={
      () =>
        navigation.navigate('Detail', { productId: product._id })
    }>
  <Image source={{ uri: product.images.alls[0] }}
  style={styles.image}>
    </Image>
    <Text style={styles.detail}>{product.name}</Text>
    <Text style={styles.cost}>{`${setCostValue2Decimals(product.cost.value)} ${product.cost.currency}`}</Text>
  </TouchableOpacity>
     </View>

   ))}
  </ScrollView>
  </View>

  );
}

function mapStateToProps ({ products: { allProducts } }) {
  return { allProducts };
}

export default connect(mapStateToProps, null)(NewProducts);
