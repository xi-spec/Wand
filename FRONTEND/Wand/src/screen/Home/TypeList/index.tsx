import React, { useEffect } from 'react';
import {
  Text,
  ImageBackground,
  View

} from 'react-native';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import {
  loadProductsType
} from '../../../redux/actions/productActions';
import styles from './styles';
import { NavigationScreenProp, NavigationRoute } from 'react-navigation';
import { Params } from '../../../models';
import { ProductType } from '../../../models/initiaState';
import * as Animatable from 'react-native-animatable';
import Swiper from 'react-native-swiper/src';
import { TouchableWithoutFeedback } from 'react-native-gesture-handler';

function TypeList ({ actions, productsType, navigation }:{
  navigation: NavigationScreenProp<NavigationRoute<Params>, Params>, actions:any, productsType:ProductType[]}) {
  useEffect(() => {
    actions.loadProductsType();
  }, []);
  return (

  <View style={{ flex: 1 }}>
    <View style={styles.container}>
   <Swiper
       bounces={false}
       showsButtons={false}
       removeClippedSubviews={false}
       key={Math.random() * 99}
       autoplay={true}
       autoplayDirection={true}
       paginationStyle={styles.paginationStyle}
       dotStyle={styles.dotStyle}
       activeDotStyle={styles.activeDotStyle}>
         {productsType && productsType.map(type => (
           <TouchableWithoutFeedback
           testID='navigation'
           onPress={() => navigation.navigate('List', { typeName: type.name })}
           key={Math.random() * 99}>
           <ImageBackground
           source={{ uri: type.image }}
           style={styles.image}>
             <View style={styles.nameContainer}>
             <Text style={styles.nameText}>{type.name.toUpperCase()}</Text>
             <Animatable.Text duration={800} animation="pulse" easing="ease-in" iterationCount="infinite" style={styles.seeAlltext}>SEE ALL</Animatable.Text>
             </View>
             </ ImageBackground>
           </TouchableWithoutFeedback>
         ))}
          </Swiper>
          <View style={styles.textContainer}>
          <Text style={styles.text}>WELCOME TO WAND SHOP</Text>
          </View>
    </View>
</View>
  );
}

function mapStateToProps ({ productsType }) {
  return { productsType };
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators({
      loadProductsType
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(TypeList)
;
