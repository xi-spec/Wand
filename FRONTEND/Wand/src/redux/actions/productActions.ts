import actionTypes from './actionTypes';
import axios from 'axios';
import SERVER from '../../constants/server';
import { Dispatch } from 'redux';
import { Product, ShoppingCart } from '../../models/initiaState';

export function loadProductsType () {
  return async (dispatch : Dispatch) => {
    const { data } = await axios.get(SERVER.URL_TYPE);
    dispatch({
      type: actionTypes.LOAD_PRODUCT_TYPE,
      data
    });
  };
}

export function loadAllProducts () {
  return async (dispatch : Dispatch) => {
    const { data } = await axios.get(SERVER.URL_PRODUCT);
    dispatch({
      type: actionTypes.LOAD_ALL_PRODUCTS,
      data
    });
  };
}

export function loadProductsList (typeName:string) {
  return {
    type: actionTypes.LOAD_PRODUCTS_LIST,
    typeName
  };
}

export function loadProductDetail (productId:string) {
  return {
    type: actionTypes.LOAD_PRODUCT_DETAIL,
    productId
  };
}

export function addProductToCart (product:Product) {
  return {
    type: actionTypes.ADD_PRODUCT_TO_CART,
    product
  };
}

export function increaseQuantityCart (productId:string) {
  return {
    type: actionTypes.INCREASE_QUANTITY_CART,
    productId
  };
}

export function decreaseQuantityCart (productId:string) {
  return {
    type: actionTypes.DECREASE_QUANTITY_CART,
    productId
  };
}

export function deleteProductCart (productId:string, quantity:number) {
  return {
    type: actionTypes.DELETE_PRODUCT_CART,
    productId,
    quantity
  };
}

export function searchProducts (name:string) {
  return {
    type: actionTypes.SEARCH_PRODUCT,
    name
  };
}

export function updateStockApi (cart:ShoppingCart) {
  return (dispatch : Dispatch) => {
    cart.cartList.forEach(async (product) => {
      const { data } = await axios.put(`${SERVER.URL_PRODUCT}/${product._id}`, { stock: product.stock });
      dispatch({
        type: actionTypes.UPDATE_STOCK_API,
        data
      });
    });
  };
}
