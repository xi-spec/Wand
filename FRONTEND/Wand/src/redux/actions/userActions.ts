import axios from 'axios';
import actionTypes from './actionTypes';
import SERVER from '../../constants/server';
import { User, Product, ShoppingCart } from '../../models/initiaState';
import { Dispatch } from 'redux';

export function addProductToFavourite (product:Product, user:User, favouriteState:boolean) {
  let newUser;
  favouriteState
    ? newUser = { ...user, favourite: [...user.favourite, product] }
    : newUser = { ...user, favourite: user.favourite.filter(elem => elem._id !== product._id) };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, newUser);
    dispatch({
      type: actionTypes.ADD_PRODUCT_TO_FAVOURITE,
      data
    });
  };
}

export function register (email:string, name:string, password:string) {
  const query = { email, name, password };
  let actionObj;
  return async (dispatch:Dispatch) => {
    try {
      const { data, status } = await axios.post(SERVER.URL_REGISTER, query);
      actionObj = {
        type: actionTypes.REGISTER,
        data,
        status
      };
    } catch (error) {
      actionObj = {
        type: actionTypes.REGISTER_ERROR
      };
    }
    dispatch(
      actionObj
    );
  };
}

export function login (email:string, password:string) {
  const query = { email, password };

  return async (dispatch:Dispatch) => {
    let actionObj;
    try {
      const { data } = await axios.post(SERVER.URL_LOGIN, query);
      actionObj = {
        type: actionTypes.LOGIN,
        data

      };
    } catch (error) {
      actionObj = {
        type: actionTypes.LOGIN_ERROR
      };
    }

    dispatch(actionObj);
  };
}

export function logout (email:string, password:string) {
  const query = { email, password };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.post(SERVER.URL_LOGOUT, query);

    dispatch({
      type: actionTypes.LOGOUT,
      data
    });
  };
}

export function cleanIsExistStatus () {
  return {
    type: actionTypes.CLEAN_ISEXIST_STATUS
  };
}

export function cleanLoginError () {
  return {
    type: actionTypes.CLEAN_LOGIN_ERROR
  };
}

export function cleanRegisted () {
  return {
    type: actionTypes.CLEAN_REGISTED
  };
}

export function removeFavourite (product:Product, user:User) {
  const newUser = { ...user, favourite: user.favourite.filter(elem => elem._id !== product._id) };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, newUser);
    dispatch({
      type: actionTypes.REMOVE_FAVOURITE,
      data
    });
  };
}

export function updateAddress (street:string,
  city:string, country:string, postcode:string, user:User) {
  const query = { street, city, country, postcode };
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { address: query });
    dispatch({
      type: actionTypes.UPDATE_USER,
      data
    });
  };
}

export function updateCard (cardNumber:string, securtyCode:string, cardholder:string, expiryDate:string, user:User) {
  const query = { cardNumber, securtyCode, cardholder, expiryDate };
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { card: query });
    dispatch({
      type: actionTypes.UPDATE_USER,
      data
    });
  };
}

export function paymentSuccessfuly (user:User, cart:ShoppingCart) {
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { orderHistory: [...user.orderHistory, cart] });
    dispatch({
      type: actionTypes.PAYMENT_SUCCSSFULY,
      data
    });
  };
}
