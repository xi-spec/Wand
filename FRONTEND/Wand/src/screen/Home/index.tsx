import React, { useEffect } from 'react';
import TypeList from './TypeList';
import NewProducts from './NewProducts';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { loadAllProducts } from '../../redux/actions/productActions';
import { ScrollView } from 'react-native-gesture-handler';
import { NavigationScreenProp, NavigationRoute } from 'react-navigation';
import { Params } from '../../models';
import { Products } from '../../models/initiaState';

function Home ({ actions, products, navigation }:{
   navigation: NavigationScreenProp<NavigationRoute<Params>, Params>, actions:any, products:Products}) {
  useEffect(() => {
    actions.loadAllProducts();
  }, [products.allProducts.length]);

  return (

    <ScrollView
    bounces={false}
    showsVerticalScrollIndicator={false}>
    <TypeList navigation={navigation}/>
    <NewProducts navigation={navigation}/>
    </ScrollView>
  );
}

function mapStateToProps ({ products }) {
  return { products };
}
function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators({
      loadAllProducts
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Home)
;
