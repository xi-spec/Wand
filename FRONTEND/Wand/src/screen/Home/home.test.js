import React from 'react';
import Home from './index';
import configureStore from 'redux-mock-store';
import * as actions from '../../redux/actions/productActions';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { render } from '@testing-library/react-native';

jest.mock('@gorhom/bottom-sheet', () => '');
jest.mock('react-native-swiper/src', () => '');

describe('Given a Home component', () => {
  const mockStore = configureStore([thunk]);
  const store = mockStore({
    products: {
      allProducts: [{ images: { alls: ['url'] }, cost: { value: 1, currency: 'euro' } }]
    }
  });

  describe('When is rendered', () => {
    test('Then is rendered correct', () => {
      const tree = render(<Provider store={store}><Home /></Provider>);
      expect(tree).toMatchSnapshot();
    });

    test('Then actions.loadAllproducts is invoked', () => {
      jest.spyOn(actions, 'loadAllProducts').mockReturnValueOnce({ type: '' });
      render(<Provider store={store}><Home /></Provider>);
      expect(actions.loadAllProducts).toHaveBeenCalled();
    });
  });
});
