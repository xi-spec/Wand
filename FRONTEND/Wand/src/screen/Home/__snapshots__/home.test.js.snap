// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Home component When is rendered Then is rendered correct 1`] = `
<RCTScrollView
  bounces={false}
  collapsable={false}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  showsVerticalScrollIndicator={false}
>
  <View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "height": 1227.28,
            "overflow": "hidden",
            "paddingBottom": 10,
          }
        }
      >
        <
          activeDotStyle={
            Object {
              "backgroundColor": "#2E3192",
            }
          }
          autoplay={true}
          autoplayDirection={true}
          bounces={false}
          dotStyle={
            Object {
              "backgroundColor": "#E6E6E6",
              "opacity": 0.4,
            }
          }
          paginationStyle={
            Object {
              "bottom": 6,
            }
          }
          removeClippedSubviews={false}
          showsButtons={false}
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2E3192",
              "height": 66.7,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontFamily": "RalewayRegular",
                "fontSize": 18,
                "margin": 5,
                "textAlign": "center",
              }
            }
          >
            WELCOME TO WAND SHOP
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "margin": 10,
          "paddingBottom": 50,
          "paddingTop": 30,
        }
      }
    >
      <Text
        style={
          Object {
            "fontFamily": "RalewayRegular",
            "fontSize": 18,
            "margin": 10,
          }
        }
      >
        New arrivals
      </Text>
      <RCTScrollView
        bounces={false}
        horizontal={true}
        showsHorizontalScrollIndicator={false}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</RCTScrollView>
`;
